base: Allow faction-based docking rights (12.10.20)

dd84ai opened the pull request:
feat(baseplugin): adding new docking permissions based on IFF.

adding new commands /base addfac, /base rmfac, /base lstfac, /base myfac
they make possible to add as friendly or hostile a whole faction by their IFF
this changes their docking permissions and friendly/hostile attitude to base

dsyalex commented on Plugins/Public/base_plugin/PlayerBase.cpp:
```diff
@@ -283,6 +283,13 @@ void PlayerBase::Load()
 					}
+					else if (ini.is_value("faction_ally_tag"))
+					{
+					wstring tag;
```
broken indentation

dd84ai replied:
Done

dsyalex commented on Plugins/Public/base_plugin/PlayerCommands.cpp:
```diff
@@ -451,6 +451,119 @@ namespace PlayerCommands
+	{
+		uint aff = GetAffliationFromClient(client);
+
+		wstring theaffiliation = (HkGetWStringFromIDS(Reputation::get_name(aff))).c_str();
```
extra brackets

dd84ai replied:
After searching for a bit, I think that redundant brackets (parenthesis) does nothing for the resulting code
https://www.quora.com/Extra-parentheses-is-have-any-effect-on-C-C++-program-or-a-time-of-compilation
Except for some unique situations.

Fixed anyway to -> wstring theaffiliation = HkGetWStringFromIDS(Reputation::get_name(aff)).c_str();

dsyalex commented on Plugins/Public/base_plugin/PlayerCommands.cpp:
```diff
@@ -451,6 +451,119 @@ namespace PlayerCommands
+		std::wstring widestr = std::wstring(theaffiliation.begin(), theaffiliation.end());
+		const wchar_t* widecstr = widestr.c_str();
+
+		PrintUserCmdText(client, L"Ship aff: %d, %s", aff, widecstr);
```
```suggestion
		PrintUserCmdText(client, L"Ship affiliation: %d, %s", aff, widecstr);

```

dd84ai replied:
Ok

dsyalex commented on Plugins/Public/base_plugin/PlayerCommands.cpp:
```diff
@@ -451,6 +451,119 @@ namespace PlayerCommands
+		wstring theaffiliation = (HkGetWStringFromIDS(Reputation::get_name(aff))).c_str();
+
+		std::wstring widestr = std::wstring(theaffiliation.begin(), theaffiliation.end());
+		const wchar_t* widecstr = widestr.c_str();
```
just do this in the PrintUserCmdText call

dd84ai replied:
It works.

dsyalex commented on Plugins/Public/base_plugin/PlayerBase.cpp:
```diff
@@ -536,14 +547,21 @@ float PlayerBase::GetAttitudeTowardsClient(uint client)
 
+	uint playeraff = GetAffliationFromClient(client);
+	// Make base friendly if player is on the friendly faction list.
+	if ((ally_factions).find(playeraff) != (ally_factions).end())
```
extra brackets

dd84ai replied:
fixed

dsyalex commented on Plugins/Public/base_plugin/PlayerBase.cpp:
```diff
@@ -536,14 +547,21 @@ float PlayerBase::GetAttitudeTowardsClient(uint client)
 			return 1.0;
 		}
-	}
+	}	
```
???

dd84ai replied:
What? Changed by accident. Not sure how to remove it.

dd84ai replied:
The resulting change changes nothing anyway.

dsyalex commented on Plugins/Public/base_plugin/Main.cpp:
```diff
@@ -1078,6 +1122,22 @@ static bool IsDockingAllowed(PlayerBase *base, uint client)
+
+	//Allow dock if player is on the friendly faction list.
+	uint playeraff = GetAffliationFromClient(client);
+	if ((base->ally_factions).find(playeraff) != (base->ally_factions).end())
```
extra brackets

dd84ai replied:
done

dsyalex commented on Plugins/Public/base_plugin/Main.cpp:
```diff
@@ -976,6 +987,37 @@ bool UserCmd_Process(uint client, const wstring &args)
+	else if (args.find(L"/base addfac") == 0)
+	{
+		returncode = SKIPPLUGINS_NOFUNCTIONCALL;
+		PrintUserCmdText(client, L"Proceeding...");
```
???

dd84ai replied:
It looks normal, I see nothing unusual. Perhaps do you mean not necessary console msg? Removed.

dsyalex commented on Plugins/Public/base_plugin/PlayerCommands.cpp:
```diff
@@ -451,6 +451,119 @@ namespace PlayerCommands
+			return;
+		}
+
+		if ((base->ally_factions).find(tag) == (base->ally_factions).end())
```
extra brackets

dd84ai replied:
done

dsyalex commented on Plugins/Public/base_plugin/PlayerCommands.cpp:
```diff
@@ -451,6 +451,119 @@ namespace PlayerCommands
+		set<uint>::iterator it;
+		for (it = base->ally_factions.begin(); it != base->ally_factions.end(); ++it)
+		{
+			wstring theaffiliation = (HkGetWStringFromIDS(Reputation::get_name(*it))).c_str();
```
extra brackets

dd84ai replied:
Done

dsyalex reviewed (changes requested):
I'm extremely hesitant to allow any extra complexity around POB docking rules.

Please squash your commits.

dd84ai commented:
> I'm extremely hesitant to allow any extra complexity around POB docking rules.
> 
> Please squash your commits.

This is why I added it as a new friendly list, better than adding new defensemods ;)
Tested more or less in local server, worked fine. Hmm I didn't make feature to show list of faction IDs in-game though, I'll try to do it perhaps, while I still have some time.
This feature was asked from me by other people, quite desired one. After some thinking I would not mind having it too ;)

dd84ai commented:
Added desires feature, /aff all that shows affiliation IDs for all factions.
because CreateID is working as well, as I get the thing with converting between regular data and the ones which flhook uses usually.

dd84ai commented:
Missed adding descriptions to help command, I ll add later

dd84ai commented:
Added

dd84ai commented:
https://discoverygc.com/forums/showthread.php?tid=183356
people opinion about it

dd84ai commented:
Removed function for hash written on its own from your Python inspiration
because there is already MakeID in flhook which does the same.
and wrapped my trash into class. It looks slightly more compact in this way.

dsyalex commented on Plugins/Public/base_plugin/PlayerCommands.cpp:
```diff
@@ -451,6 +456,235 @@ namespace PlayerCommands
+				for (map<string, uint>::iterator iter = factions.begin(); iter != factions.end(); iter++)
+				{
+					string factionnickname = iter->first;
+					uint ID = MakeId(factionnickname.c_str());
```
Can we document what this does? Nothing else appears to use it. I assume this is like CreateID but for faction hashes?

dd84ai replied:
createIDTable = None
def FLFactionHash(nickName):
        FLFACTIONHASH_POLYNOMIAL = 0x1021

        # Build the crc lookup table if it hasn't been created
        global createIDTable
        if createIDTable is None:
                createIDTable = []
                for i in range(0, 2**8):
                        x = i << 8
                        for bit in range(0, 8):
                                x <<= 1
                                if x & 0x10000:
                                        x = (x ^ FLFACTIONHASH_POLYNOMIAL) & 0xFFFF
                        createIDTable.append(x)

        # Calculate the hash.
        hash = 0xFFFF
        for c in bytearray(nickName.lower().encode('utf-8')):
                hash = (hash >> 8) ^ createIDTable[(hash & 0xFF) ^ c]
        return hash

if __name__ == "__main__":
        import sys
        for n in sys.argv[1:]:
                h = FLFactionHash(n)
                
print(n, h, '0x{:04x}'.format(h))
alex@alex-laptop:~/game-repository (Development)$ python3 SERVICE/alexscripts/createfactionid.py fc_freelancer
fc_freelancer 4169 0x1049

Do you remember you gave me this function? Later I rewrote it to C++..... and a long time later after that Dartes told me about an already existing function in Flhook which does the same ;) He got the info from Laz. Basically MakeID is this function above, but in built in FLhook, same results.

dd84ai replied:
Provided comment to code
//MakeID function (in built in Flhook) is the same as mentioned here in C# to CreateFactionID https://github.com/DiscoveryGC/FLHook/blob/master/Plugins/Public/playercntl_plugin/setup_src/FLUtility.cs

dsyalex commented on Plugins/Public/base_plugin/PlayerCommands.cpp:
```diff
@@ -451,6 +456,235 @@ namespace PlayerCommands
+				{
+					string factionnickname = iter->first;
+					uint ID = MakeId(factionnickname.c_str());
+					wstring factionname = GetFactioName(ID);;
```
double semicolon, also function name has a typo - `GetFactionName`

dd84ai replied:
Got it.

dsyalex commented on Plugins/Public/base_plugin/PlayerCommands.cpp:
```diff
@@ -451,6 +456,235 @@ namespace PlayerCommands
+			}
+			catch (exception e)
+			{
+				return L"No Affiliation";
```
Unknown Reputation?

dd84ai replied:
Fixed

dsyalex commented on Plugins/Public/base_plugin/PlayerCommands.cpp:
```diff
@@ -451,6 +456,235 @@ namespace PlayerCommands
+			LoadAffList();
+		}
+
+		void FindAndPrintOneAffiliation(uint client, uint smth)
```
smth?

dd84ai replied:
let's call it 'AffiliationID' then. (Which was created by hash function)

dsyalex commented on Plugins/Public/base_plugin/PlayerCommands.cpp:
```diff
@@ -451,6 +456,235 @@ namespace PlayerCommands
+			std::list<Affiliations::AffCell>::iterator found;
+			found = std::find_if(AffList.begin(), AffList.end(), std::bind(IDComparision, std::placeholders::_1, smth));
+			if (found != AffList.end())
+				PrintUserCmdText(client, L"Aff ID: %u, %s, %s", found->id, (found->nickname).c_str(), (found->factionname).c_str());
```
Aff? Let's not send "Aff" to players

dd84ai replied:
Lets call it IFF ID, people approve.

dsyalex commented on Plugins/Public/base_plugin/PlayerCommands.cpp:
```diff
@@ -451,6 +456,235 @@ namespace PlayerCommands
+			if (found != AffList.end())
+				PrintUserCmdText(client, L"Aff ID: %u, %s, %s", found->id, (found->nickname).c_str(), (found->factionname).c_str());
+			else
+				PrintUserCmdText(client, L"Aff ID: %u, Unknown, Unknown", smth);
```
see above

dd84ai replied:
see above

dsyalex commented on Plugins/Public/base_plugin/PlayerCommands.cpp:
```diff
@@ -451,6 +456,235 @@ namespace PlayerCommands
+		{
+			for (list<Affiliations::AffCell>::iterator iter = AffList.begin(); iter != AffList.end(); iter++)
+			{
+				PrintUserCmdText(client, L"Aff ID: %d, %s, %s", iter->id, (iter->nickname).c_str(), (iter->factionname).c_str());
```
see above

dd84ai replied:
see above

dsyalex reviewed (changes requested):
Please write a commit message

dd84ai commented:
Resolved everything

dsyalex reviewed (changes requested):
requires rebase

dd84ai commented:
Performed tests:
1) /base myfac works well. it shows expected output
2) Attempt to use addfac, rmfac and lstfac failed because I am not logged as base admin
3) logged as admin and attempt to add incorrect faction. Denied.
4) added two correct factions among whom my current faction, turned defensemode as hostile 2, and attempted to redock. I was green and succeeded
5) Reloaded plugin to see if it will persist through reloading. Succeded.
6) Deleted correct faction from list, succeded
7) Deleted not correct faction from list, not succeded
8) Deleted all factions and tried to redock
